main/src/main/scala/sbt/Defaults.scala

```diff
@@ -2571,21 +2556,11 @@ object Defaults extends BuildCommon {
       val contents = AnalysisContents.create(analysisResult.analysis(), analysisResult.setup())
       store.set(contents)
       Def.declareOutput(analysisOut)
-      val dir = ci.options.classesDirectory.toFile()
-      val mappings = Path
-        .allSubpaths(dir)
-        .filter(_._1.isFile())
-        .map { case (p, path) =>
-          val vf = c.toVirtualFile(p.toPath())
-          (vf: HashedVirtualFileRef) -> path
-        }
-        .toSeq
-      // inlined to avoid caching mappings
-      val pkgConfig = Pkg.Configuration(mappings, artifactPath.value, packageOptions.value)
-      val out = Pkg(pkgConfig, c, s.log, Pkg.timeFromConfiguration(pkgConfig))
-      s.log.debug(s"wrote $out")
-      Def.declareOutput(out)
-      analysisResult.hasModified() -> (out: HashedVirtualFileRef)
+      val dir = ci.options.classesDirectory
+      val vfDir = c.toVirtualFile(dir)
+      val packedDir = Def.declareOutputDirectory(vfDir)
+      s.log.debug(s"wrote $vfDir")
+      (analysisResult.hasModified(), vfDir: VirtualFileRef, packedDir: HashedVirtualFileRef)
```

Comment:
See the "beware the hermeticity breakage" section in https://eed3si9n.com/sudori-part4/.

> The `compile` task example reminds me of a potential pitfall with `Def.declareOutputDirectory` when used with `Def.cachedTask(...)`, which is that if we’re not careful, we could end up breaking the hermeticity. Specifically in this case, we could end up reusing an old, incorrect cache.
>
> This is because `VirtualFileRef` that represents the directory name will not contain the hash information of the content, which does not contain enough information to invalidate the downstream tasks. For example, if a cached task foo produces a directory and passes it to another cached task bar, passing `VirtualFileRef` alone will not be a sufficient cache.
>
> To workaround this, `Def.declareOutputDirectory` returns `VirtualFile` for the synthetic zip file, which does contain the hash, and should be able to invalidate the downstream.

Reply:
Thanks, I do understand now but I need to think more about it.